### PR TITLE
BUG: raise on non-hashable Index name, closes #29069

### LINKED ISSUE #29069: Index.name is not validatied to be Hashable
We require that Series.name be hashable. We don't do the same for Index.name. Should we?

```python
In [4]: pd.Index([], name=[])
Out[4]: Index([], dtype='object', name=[])
```

```python
In [5]: pd.Series([], name=[])
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-2c73ddde103e> in <module>
----> 1 pd.Series([], name=[])

~/sandbox/pandas/pandas/core/series.py in __init__(self, data, index, dtype, name, copy, fastpath)
    326         generic.NDFrame.__init__(self, data, fastpath=True)
    327
--> 328         self.name = name
    329         self._set_axis(0, index, fastpath=True)
    330

~/sandbox/pandas/pandas/core/generic.py in __setattr__(self, name, value)
   5257             object.__setattr__(self, name, value)
   5258         elif name in self._metadata:
-> 5259             object.__setattr__(self, name, value)
   5260         else:
   5261             try:

~/sandbox/pandas/pandas/core/series.py in name(self, value)
    468     def name(self, value):
    469         if value is not None and not is_hashable(value):
--> 470             raise TypeError("Series.name must be a hashable type")
    471         object.__setattr__(self, "_name", value)
    472

TypeError: Series.name must be a hashable type
```

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #29069
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by TomAugspurger** [comment]
To close https://github.com/pandas-dev/pandas/issues/29069, Index.name should probably be a settable property, and the setter should validate that the name is hashable.

Any perf concerns here? Things should be OK right?

**Comment by jbrockmendel** [comment]
huh, looks like i got confused by `names`, which is a property.  will update.

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -5464,3 +5463,17 @@ def default_index(n):
     from pandas.core.indexes.range import RangeIndex
 
     return RangeIndex(0, n, name=None)
+
+
+def maybe_extract_name(name, obj):
```
Comment (jreback) [inline pandas/core/indexes/base.py]: i think this shares with series yes?

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — COMMENTED** [review]
can you also add a whatsnew note

**Comment by jbrockmendel** [comment]
Are we OK with being able to set MultiIndex.name?  When I tried disallowing it, some tests broke

**Comment by jbrockmendel** [comment]
xref #11979 should we try to address that here while were at it?

File: pandas/core/indexes/base.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/indexes/base.py]: yep

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -5464,3 +5472,19 @@ def default_index(n):
...
+    If no name is passed, then extract it from data, validating hashability.
+    """
+    if name is None and isinstance(obj, (Index, ABCSeries)):
+        # Note we don't just check for "name" attribute since that would
+        #  pick up e.g. dtype.name
+        name = obj.name
+
+    # GH#29069
+    if not is_hashable(name):
+        raise TypeError(f"{cls_name}.name must be a hashable type")
```
Comment (jreback) [inline pandas/core/indexes/base.py]: can you use type(obj).__name__ here?

File: pandas/core/indexes/base.py [hunk 3]
```diff
@@ -5464,3 +5472,19 @@ def default_index(n):
     from pandas.core.indexes.range import RangeIndex
 
     return RangeIndex(0, n, name=None)
+
+
+def maybe_extract_name(name, obj, cls_name="Index"):
```
Comment (jreback) [inline pandas/core/indexes/base.py]: can you type and add a doc-string

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. some changes

**Comment by jbrockmendel** [comment]
I think #19171 becomes easier after this.

File: pandas/core/indexes/base.py [hunk 4]
```diff
@@ -1209,6 +1208,15 @@ def to_frame(self, index=True, name=None):
     # --------------------------------------------------------------------
     # Name-Centric Methods
 
+    @property
+    def name(self):
+        return self._name
+
+    @name.setter
+    def name(self, value):
+        maybe_extract_name(value, None, type(self))
```
Comment (jreback) [inline pandas/core/indexes/base.py]: in series.py you can change name.setter to use this I think (ok for followon)

File: pandas/core/indexes/base.py [hunk 5]
```diff
@@ -5462,3 +5470,19 @@ def default_index(n):
     from pandas.core.indexes.range import RangeIndex
 
     return RangeIndex(0, n, name=None)
+
+
+def maybe_extract_name(name, obj, cls) -> Optional[Hashable]:
```
Comment (jreback) [inline pandas/core/indexes/base.py]: see my comment above, can be more general so need to relocate i think as this is common between series/index)

**Review by jreback — APPROVED** [review]
comments for a followon

**Comment by jreback** [comment]
thanks

**Pull request merged**